processing/src/main/java/org/apache/druid/segment/join/JoinFilterAnalyzer.java

```diff
@@ -0,0 +1,743 @@
+              correlationCache
+          );
+          if (!rewritten.isCanPushDown()) {
+            return JoinFilterAnalysis.createNoPushdownFilterAnalysis(orFilter);
+          } else {
+            newFilters.add(rewritten.getPushdownFilter());
+          }
+        } else {
+          return JoinFilterAnalysis.createNoPushdownFilterAnalysis(orFilter);
+        }
+      } else {
+        newFilters.add(filter);
+      }
+    }
+
+    return new JoinFilterAnalysis(
+        true,
+        retainRhs,
+        orFilter,
+        new OrFilter(newFilters),
+        null
+    );
+  }
+
+  /**
+   * Rewrites a selector filter on a join table into an IN filter on the base table.
+   *
+   * @param baseAdapter      The adapter for the join
+   * @param selectorFilter   SelectorFilter to be rewritten
+   * @param prefixes         Map of join table prefixes to clauses
+   * @param equiconditions   Map of equiconditions
+   * @param correlationCache Cache of column correlation analyses
+   *
+   * @return A JoinFilterAnalysis that indicates how to handle the potentially rewritten filter
+   */
+  private static JoinFilterAnalysis rewriteSelectorFilter(
+      HashJoinSegmentStorageAdapter baseAdapter,
+      SelectorFilter selectorFilter,
+      Map<String, JoinableClause> prefixes,
+      Map<String, Expr> equiconditions,
+      Map<String, List<JoinFilterColumnCorrelationAnalysis>> correlationCache
+  )
+  {
+    String filteringColumn = selectorFilter.getDimension();
+    for (Map.Entry<String, JoinableClause> prefixAndClause : prefixes.entrySet()) {
+      if (filteringColumn.startsWith(prefixAndClause.getKey())) {
+        List<JoinFilterColumnCorrelationAnalysis> correlations = correlationCache.computeIfAbsent(
+            prefixAndClause.getKey(),
+            p -> findCorrelatedBaseTableColumns(
+                baseAdapter,
+                p,
+                prefixes.get(p),
+                equiconditions
+            )
+        );
+
+        if (correlations == null) {
+          return JoinFilterAnalysis.createNoPushdownFilterAnalysis(selectorFilter);
+        }
+
+        List<Filter> newFilters = new ArrayList<>();
+        List<VirtualColumn> pushdownVirtualColumns = new ArrayList<>();
+
+        for (JoinFilterColumnCorrelationAnalysis correlationAnalysis : correlations) {
+          if (correlationAnalysis.supportsPushDown()) {
+            List<String> correlatedValues = getCorrelatedValuesForPushDown(
+                selectorFilter.getDimension(),
+                selectorFilter.getValue(),
+                correlationAnalysis.getJoinColumn(),
+                prefixAndClause.getValue()
+            );
+
+            if (correlatedValues == null) {
+              return JoinFilterAnalysis.createNoPushdownFilterAnalysis(selectorFilter);
+            }
+
+            for (String correlatedBaseColumn : correlationAnalysis.getBaseColumns()) {
+              InFilter rewrittenFilter = (InFilter) new InDimFilter(
```

Comment:
Is there a reason for this cast? Seems like we could let it go in to `newFilters` as an uncast Filter.

Reply:
Hm, it must have been a relic from an earlier version of the code, adjusted as suggested